data/transactions/logic/debugger.go

```diff
@@ -34,14 +34,157 @@ import (
+	// │  │ > AfterAppEval                           │  │
+	// │  └──────────────────────────────────────────┘  │
+	// |    ⁞  ⁞  ⁞  ⁞  ⁞  ⁞  ⁞  ⁞  ⁞  ⁞  ⁞  ⁞  ⁞  ⁞    │
+	// │                                                │
+	// │ > AfterTxn                                     │
+	// └────────────────────────────────────────────────┘
+
+}
+
+type debuggerBeforeTxnHook interface {
+	// BeforeTxn is called before the transaction is executed
+	BeforeTxn(ep *EvalParams, groupIndex int) error
+}
+
+func CallBeforeTxnHookIfItExists(dh DebuggerHook, ep *EvalParams, groupIndex int) error {
+	if dhWithBeforeTxnHook, ok := dh.(debuggerBeforeTxnHook); ok {
+		return dhWithBeforeTxnHook.BeforeTxn(ep, groupIndex)
+	}
+	return nil
+}
+
+type debuggerAfterTxnHook interface {
+	// AfterTxn is called after the transaction has been executed.
+	// groupIndex refers to the index of the transaction in the transaction group that was just executed.
+	AfterTxn(ep *EvalParams, groupIndex int) error
+}
+
+func CallAfterTxnHookIfItExists(dh DebuggerHook, ep *EvalParams, groupIndex int) error {
+	if dhWithAfterTxnHook, ok := dh.(debuggerAfterTxnHook); ok {
+		return dhWithAfterTxnHook.AfterTxn(ep, groupIndex)
+	}
+	return nil
+}
+
+type debuggerBeforeAppEvalHook interface {
+	// BeforeAppEval is called before the app is evaluated.
+	// This hook is similar to BeforeTxn, but includes debug state information instead of eval params.
+	BeforeAppEval(state *DebugState) error
+}
+
+func callBeforeAppEvalHookIfItExists(dh DebuggerHook, state *DebugState) error {
+	if dhWithBeforeAppEvalHook, ok := dh.(debuggerBeforeAppEvalHook); ok {
+		return dhWithBeforeAppEvalHook.BeforeAppEval(state)
+	}
+	return nil
+}
+
+type debuggerAfterAppEvalHook interface {
+	// AfterAppEval is called after the app has been evaluated.
+	AfterAppEval(state *DebugState) error
+}
+
+func callAfterAppEvalHookIfItExists(dh DebuggerHook, state *DebugState) error {
+	if dhWithAfterAppEvalHook, ok := dh.(debuggerAfterAppEvalHook); ok {
+		return dhWithAfterAppEvalHook.AfterAppEval(state)
+	}
+	return nil
+}
+
+type debuggerBeforeTealOpHook interface {
+	// BeforeTealOp is called before the op is evaluated
+	BeforeTealOp(state *DebugState) error
+}
+
+func callBeforeTealOpHookIfItExists(dh DebuggerHook, state *DebugState) error {
+	if dhWithBeforeTealOpHook, ok := dh.(debuggerBeforeTealOpHook); ok {
+		return dhWithBeforeTealOpHook.BeforeTealOp(state)
+	}
+	return nil
+}
+
+type debuggerAfterTealOpHook interface {
+	// AfterTealOp is called after the op has been evaluated
+	AfterTealOp(state *DebugState) error
+}
+
+func callAfterTealOpHookIfItExists(dh DebuggerHook, state *DebugState) error {
+	if dhWithAfterTealOpHook, ok := dh.(debuggerAfterTealOpHook); ok {
+		return dhWithAfterTealOpHook.AfterTealOp(state)
+	}
+	return nil
+}
+
+type debuggerBeforeInnerTxnHook interface {
+	// BeforeInnerTxn is called before the inner transaction is executed
+	BeforeInnerTxn(ep *EvalParams) error
```

Comment:
Shouldn't group index be passed here as well, otherwise there's no easy to way to tell which txn is being executed?

Reply:
No because this hook is meant to indicate that an inner transaction is about to be executed (think something equivalent to a non-existent `BeforeTxnGroup`).

Instead, the `BeforeTxn` hook is _supposed to be_ re-used before each transaction submitted within an `itxn_submit`. However, it seems I lost track of that when implementing the hooks within the evaluation flow and so it's not the case currently. Once I fix that, you can use `BeforeTxn` to get each inner transaction's group index.

Reply:
Ok, that makes sense then. I'd suggest clarifying this in the docs so that it's clear for future readers too. And perhaps renaming to `BeforeInnerTxnGroup`?

Reply:
Good idea! I'll rename it and update the docs.